### ISSUE TITLE
HashMap starts the search even if the map is empty

### ISSUE THREAD

**Issue by pmarcelll** [rust-lang/rust#38880]
It was mentioned in the latest [Ferris Makes Emulators episode](https://youtu.be/d-dm4z5fJ1U?t=1h14m29s) that `contains` showed up in the profiling results, although a lot of the times the map itself was empty.
I think the extra checking is worth it before starting the search and that way the hashing is also skipped.

**Comment by Manishearth** [comment]
`contains()` calls [`contains_key()`](https://github.com/rust-lang/rust/blob/0728b710347aa80dfc05de29bfed4ed5fc023818/src/libstd/collections/hash/map.rs#L1106) which calls [`search()`](https://github.com/rust-lang/rust/blob/0728b710347aa80dfc05de29bfed4ed5fc023818/src/libstd/collections/hash/map.rs#L506) and finally [`search_hashed()`](https://github.com/rust-lang/rust/blob/0728b710347aa80dfc05de29bfed4ed5fc023818/src/libstd/collections/hash/map.rs#L363)


That function checks for zero capacity the first thing it does. The only work that is done between the `contains()` call and `search_hashed()` is computing the hash.

Now, if the map was empty (but had nonzero capacity) we'd have a problem. `search_hashed` can't bail early in that case because it still needs to produce the entry, but `contains` doesn't care about that.

In Ferris' example, were hashmaps with nonzero capacity but zero length used? You can make that happen with `HashMap::with_capacity(foo)`, but by default hashmaps are zero capacity (no allocation till you insert things into them). If not, this probably isn't the problem.

Bailing out on `size == 0` in capacity is an interesting case.

**Comment by leonardo-m** [comment]
> Small negative impact on users of regular hashmaps, large positive impact on users of preallocated hashmaps.

How much "small" is the negative impact in the regular case needs to be benchmarked.

**Comment by Manishearth** [comment]
Yes. It's probably negligible.

**Comment by Stebalien** [comment]
@Manishearth That really depends on the use case. If you have a large vector of mostly empty `HashMap`/`HashSet`s, repeatedly re-hashing the key could take a while. However, this probably isn't a common use-case so the programmer could just check the length in cases like this.

Anyways, wasn't there some talk about falling back on linear search/storage for small maps/sets? That would probably be more useful in practice.

**Comment by arthurprs** [comment]
I wanted to move the empty check up in the call stack, notably to the get* and contain methods, as entry and insert reserve at least 1 space upfront. This saves 1 branch offsetting the extra ones required by #38368

**Comment by technicalguy** [comment]
So I've had a look into this issue.

> The only work that is done between the contains() call and search_hashed() is computing the hash.

Computing the hash can be quite costly, even for integer keys on an empty `HashMap` it seems `contains()` is an order of magnitude faster with an empty check before hash computation than with one after.

> Bailing out on size == 0 in contains is an interesting idea. Small negative impact on users of regular hashmaps, large positive impact on users of preallocated hashmaps.

I think it should be possible to make an early `size == 0` check without requiring a second check afterwards (i.e. no negative impact on regular `HashMap` users with non-empty maps). Some kind of `search_hashed_no_empty_check` method could be called by a `search_no_empty_check` method, given that contains_key has already checked that it is empty.

> Anyways, wasn't there some talk about falling back on linear search/storage for small maps/sets? That would probably be more useful in practice.

This works well for small collections (e.g. integer keys up to about 16 items), however the problem is that different types require different amounts of time to compute the hash, and and also different amounts of time to compute equality. This is not necessarily consistent between values (e.g. strings or lists).

I think I can implement this for types with simple equality checks, but it cannot be used in general. (The empty check can be used in general, so I propose we implement both.)

**Comment by lucab** [comment]
Given that #48035 landed, is there anything else pending from this ticket?

**Comment by technicalguy** [comment]
For this issue

> HashMap starts the search even if the map is empty

there is nothing left to do.

For the linear search

> Anyways, wasn't there some talk about falling back on linear search/storage for small maps/sets? That would probably be more useful in practice.

I don't know if it is possible: The time to compute equality and the time to compute a hash varies with different types (these two time determine the threshold at which it becomes faster to hash than to linear search). Not only that, but moreover, it could vary with different architectures. This means that determining the threshold is probably unfeasible in any kind of generality.

(You might be able to say that for a particular processor architecture, when the entire map is in the L1 cache, and your keys are 32-bit integers, then it is faster to do a linear search if the size of your map is up to 27 entries (for example), and after that, it is faster to hash the integer key and do a regular lookup. However, if any of these assumptions change then it might not be possible to guarantee a faster lookup for a map with 27 entries. The one feature that you really want from your HashMap is that the lookup time is bounded (with a certain probability) and it's very hard to do a linear search and guarantee that you will be at least as fast as a hash-lookup.)

The only thing that could certainly be implemented is if the hashmap contains only one element, then just do a single comparison rather than a hash lookup and then the comparison. i.e. hashmaps with size 1 don't need to compute the hash.

**Comment by pmarcelll** [comment]
I didn't realize this issue wasn't closed by #48035, and I also think this should be closed since it would be more appropriate to open a new issue for future optimizations.
